implement event handler hooks

Haringat opened the pull request:
The idea is for language adapters to be able to hook into the execution of event handlers to provide more efficient execution. As the loading handlers can consume quite some time I think it would be convenient to provide more efficient execution methods. Sadly though, I know no way Java 6 could benefit from that (Futures came too late...) and I don't know whether Scala has something like async/await or something comparable. But Forgelin/Kotlin for instance could benefit from that using suspend functions. I've created a gist to show how this could look like:
https://gist.github.com/Haringat/878847744dd0d73849eafb563fd3e6b2

Haringat committed:
implement event handler hooks

marvin-roesch commented:
While I can the see motivation behind this, I don't think there would be that much use in this. If mods which to do some of their own work concurrently, they may already to so now. All your suggestion allows for is deferring the inevitable blocking to a "language" level. In your Kotlin example, I don't see how that is saving you much code at all, assuming I understand the semantics correctly. Furthermore, concurrency is virtually useless for anything interacting with MC (be it directly or through Forge), meaning the the only tangible benefit with this is for mod-specific loading of some data.

In conclusion, I don't think the additional burden this feature would be putting on language adapter maintainers (even if there's effectively only going to be 2 xD) is worth it. From a Scala perspective, there's no useful analogue to your Kotlin reference, anyways (ways of expressing this would probably require more intrusive changes to the event handling).

Haringat commented:
Thank you for your feedback.
I agree that it would just move code from the event handlers to the language adapters. However, I've just seen that forge moved on to Java 8 now. I will update the pull request later today because that way even Java can benefit from it.

Commit:
refactored event handler hooks

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=4482) <br/>All committers have signed the CLA.

Haringat commented:
@PaleoCrafter Now I've refactored the entire thing using Java 8's CompletableFutures/CompletionStages. I will update the gist accordingly within the next few days. (if required)

Commit:
fix some white-space mistakes and move error handling up one layer

liach commented on src/main/java/net/minecraftforge/fml/common/ILanguageAdapter.java:
```diff
@@ -33,6 +37,12 @@
+     * Wraps an event handler method in an {@code EventHandler} instance
+     * @param method the eventhandler method of a mod
+     */
+    public IEventHandler createEventHandler(Method method);
```
You can default this out, I think.

liach commented on src/main/java/net/minecraftforge/fml/common/FMLModContainer.java:
```diff
@@ -595,16 +593,31 @@ public void constructMod(FMLConstructionEvent event)
+                    {
+                        try
+                        {
+                            return eventHandler.handleEvent(event);
```
`return eventHandler.handleEvent(event).toCompletableFuture();`

liach commented on src/main/java/net/minecraftforge/fml/common/FMLModContainer.java:
```diff
@@ -595,16 +593,31 @@ public void constructMod(FMLConstructionEvent event)
+                    })
+                    .reduce((future1, future2) -> future1.thenComposeAsync(aVoid -> future2))
+                    .orElse(CompletableFuture.completedFuture(null))
+                    .toCompletableFuture()
```
Is this guaranteed to be supported?

Haringat replied:
Yes. Every implementation of CompletionStage MUST implement CompletionStage.toCompletableFuture(), which is meant to return a CompletableFuture that completes when the other CompletionStage implementation does and then holds the same value. So unless the implementation of CompletionStage is bugged this should work.

diesieben07 replied:
No. Implementations may throw `UnsupportedOperationException`.

Haringat replied:
@diesieben07 Sorry. You're right. I've only read the short description in the Javadoc before. I've now changed that in a way that is guaranteed to always return a CompletableFuture.

Commit:
defaulted out ILanguageAdapter.createEventHandler(). Thanks @liach

Commit:
cleaned up imports

Commit:
ensure that handleModStateEvent only works with CompletableFutures

kashike commented on src/main/java/net/minecraftforge/fml/common/FMLModContainer.java:
```diff
@@ -18,66 +18,49 @@
  */
 package net.minecraftforge.fml.common;
 
-import java.io.File;
```
don't change imports like this

kashike commented on src/main/java/net/minecraftforge/fml/common/FMLModContainer.java:
```diff
@@ -18,66 +18,49 @@
-import java.util.Set;
-
+import com.google.common.base.Strings;
+import com.google.common.collect.*;
```
no star

kashike commented on src/main/java/net/minecraftforge/fml/common/FMLModContainer.java:
```diff
@@ -595,16 +578,31 @@ public void constructMod(FMLConstructionEvent event)
+                        // safely transform the result of eventHandler.handleEvent into a CompletableFuture
+                        return new CompletableFuture<Void>().thenCompose(aVoid ->
+                        {
+                            try {
```
code style

Commit:
Revert "cleaned up imports"

This reverts commit 802d5ad68e88ed6f0639c61dfc326ce187e3f5b3.

Commit:
code cleanup

Commit:
remove wildcard import

kashike commented on src/main/java/net/minecraftforge/fml/common/FMLModContainer.java:
```diff
@@ -29,12 +29,6 @@
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.security.cert.Certificate;
-import java.util.Arrays;
```
imports, again

Commit:
reverted imports

Commit:
removed white-space changes (empty lines)

diesieben07 commented on src/main/java/net/minecraftforge/fml/common/FMLModContainer.java:
```diff
@@ -595,16 +597,34 @@ public void constructMod(FMLConstructionEvent event)
+                            catch (Throwable t)
+                            {
+                                CompletableFuture<Void> future = CompletableFuture.completedFuture(null);
+                                future.completeExceptionally(t);
```
This does absolutely nothing. Read the Javadoc.

diesieben07 commented on src/main/java/net/minecraftforge/fml/common/FMLModContainer.java:
```diff
@@ -595,16 +597,34 @@ public void constructMod(FMLConstructionEvent event)
+                    .map(eventHandler ->
+                    {
+                        // safely transform the result of eventHandler.handleEvent into a CompletableFuture
+                        return new CompletableFuture<Void>().thenCompose(aVoid ->
```
This new `CompletableFuture` will never complete. Nobody ever calls `complete` or `completeExceptionally` on it.

Haringat replied:
@diesieben07 right, I confused the behaviour of CompletableFuture.completedFuture with that of the CompletableFuture constructor.

diesieben07 reviewed (changes requested):
I really do not understand the point of this pull request.
One, it only applies to the FML lifecycle methods. Two, it does not impose any kind of ordering on the futures (actually currently it completely breaks the event bus and never calls any event methods, ever), so the code all runs in parallel still. Why have the futures then? Nobody waits on them!
And last, but certainly not least, it is a big anti-pattern to block the startup process. It already takes long enough, don't download things or do other crap like that in those methods in a blocking way unless you absolutely have to (you almost never have to).

The issues below suggest to me you have not actually tested this in any way.

Commit:
fix future handling

Haringat commented:
@diesieben07 
Well, I relied on the automated tests (which were all green).
As for the things you pointed out:

> One, it only applies to the FML lifecycle methods.

So?

> Nobody waits on them!

They are awaited by the .join() call. It didn't work becacuse of the bugs you pointed out in your change requests.

> don't download things or do other crap like that in a blocking way

It's not about blocking the main thread with big tasks but rather to do all (potentially big) things that actually need to be done at runtime in parallel so the loading is blocked, but not for too long.

I hope that I could make my point clearer now.

diesieben07 commented:
> Well, I relied on the automated tests (which were all green).

Forge's code coverage is spotty at best.

I understand why you want this now, but I still don't think it's needed. But your code looks good now 😄

mezz commented:
Thanks for contributing! It's good that you are digging into these systems.

If I read this right, this is essentially running all handlers in parallel. I'm unfamiliar with Java's futures implementation, but Minecraft is not built well for threaded processes. Will every handler still run on the main thread?
Unfortunately existing handlers will almost certainly have to remain synchronous and only a small savvy fraction of the community can benefit from the added flexibility here. Sadly if we keep things synchronous by default, then this PR's is most likely creating more overhead because of the additional futures objects.

Is that about right or am I overlooking something in the implementation here?

Also, Forge barely has any automated testing and PRs need manual testing to verify that things work as intended. Generally I will test things out as well as I can once the author has gotten things to work and tested as well as they can. If you are interested in automated testing I can connect you with some others who are interested in improving the testing here.

Haringat commented:
> Is that about right or am I overlooking something in the implementation here?

Not quite. It is still running all event handlers sequentially (the .join() call blocks for the execution) but mod author's can opt in to return a CompletionStage and do several things in parallel inside their event handlers. AFAIK implementations like CompletableFuture.runAsync() don't run on the main thread but after the mentioned join the execution continues on the main thread.
> this PR's is most likely creating more overhead because of the additional futures objects.

True, but not much because CompletableFuture.completedFuture is a pretty cheap call.
> only a small savvy fraction of the community can benefit from the added flexibility

*looking at GregoriusT*
>If you are interested in automated testing I can connect you with some others who are interested in improving the testing here.

Sadly, I am not very experienced with writing Software tests in anything that's more complicated than JavaScript, but thanks for the offer.

But you should keep in mind that what I wrote is just one use-case for the hook that came to mind. If someone finds another one we can still add that.

diesieben07 commented:
Some clarifications on the execution order with the current code:

- The event handlers are invoked inside a `thenCompose` handler. It is explicitly not defined which thread such a handler runs on. In this case it happens to be the main thread (since the future `thenCompose` is called on is already completed), but this is _undocumented behavior_. This should be changed, in my opinion.
- Event handlers for each event (preInit, init, etc.) run in sequence. If they return a future, "all bets are off", i.e. it's up to the event handler to decide where it's code runs (that's why it returns a future).
-  The event always waits for all it's futures to complete, i.e. the preInit hook does not complete until all futures from preInit event handlers have completed.

Haringat commented:
> This should be changed, in my opinion.

Actually I think it actually comes to our advantage because this way synchronous event handlers are executed on the main thread which perfectly aligns with the previous implementation.
I could wrap it using `CompletableFuture.runAsync` to have it (at least deteministically) run on a different thread (cause it uses ForkJoinPool) but that might break mods depending on having their event handlers run in the main thread (i don't know why they should in synchronous code but it would still be a breaking change)

diesieben07 commented:
> Actually I think it actually comes to our advantage because this way synchronous event handlers are executed on the main thread which perfectly aligns with the previous implementation.

Yes, due to an implementation detail, which you cannot rely on. Here's what I'd do:

Use a safe version of `toCompletableFuture`, like this:
```java
public static <T> CompletableFuture<T> toCompletableFuture(CompletionStage<T> stage) {
    try {
        return stage.toCompletableFuture();
    } catch (UnsupportedOperationException ignored) {
        CompletableFuture<T> future = new CompletableFuture<>();
        stage.whenComplete((r, x) -> {
            if (x != null)
                future.completeExceptionally(x);
            else
                future.complete(r);
        });
        return future;
    }
}
```

Then you can use `CompleteableFuture.allOf(futures).join()` to wait for all of them.

Haringat committed:
made FMLModContainer.handleModStateEvent more readable and improved return type safety of ILanguageAdapter.createEventHandler (thanks @diesieben07)

Haringat commented:
No reaction for a month...
Sorry for bumping but I would like to know whether you accept this now or not.

mezz commented:
I tried this out just now with some packs and got errors like this:

```
[21:25:00] [main/FATAL] [FML]: There is a binary discrepancy between the expected input class b$1 (b$1) and the actual class. Checksum on disk is 51416ed6, in patch 52fe6ed7. Things are probably about to go very wrong. Did you put something into the jar file?
[21:25:00] [main/FATAL] [FML]: The game is going to exit, because this is a critical error, and it is very improbable that the modded game will work, please obtain clean jar files.
```

```
Description: There was a severe problem during mod loading that has caused the game to fail

net.minecraftforge.fml.common.LoaderExceptionModCrash: Caught exception from CraftTweaker2 (crafttweaker)
Caused by: java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: object is not an instance of declaring class
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.util.concurrent.CompletableFuture.biRelay(Unknown Source)
	at java.util.concurrent.CompletableFuture.andTree(Unknown Source)
	at java.util.concurrent.CompletableFuture.allOf(Unknown Source)
	at net.minecraftforge.fml.common.FMLModContainer.handleModStateEvent(FMLModContainer.java:609)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at com.google.common.eventbus.Subscriber.invokeSubscriberMethod(Subscriber.java:91)
	at com.google.common.eventbus.Subscriber$SynchronizedSubscriber.invokeSubscriberMethod(Subscriber.java:150)
	at com.google.common.eventbus.Subscriber$1.run(Subscriber.java:76)
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:399)
	at com.google.common.eventbus.Subscriber.dispatchEvent(Subscriber.java:71)
	at com.google.common.eventbus.Dispatcher$PerThreadQueuedDispatcher.dispatch(Dispatcher.java:116)
	at com.google.common.eventbus.EventBus.post(EventBus.java:217)
	at net.minecraftforge.fml.common.LoadController.sendEventToModContainer(LoadController.java:253)
	at net.minecraftforge.fml.common.LoadController.propogateStateMessage(LoadController.java:231)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at com.google.common.eventbus.Subscriber.invokeSubscriberMethod(Subscriber.java:91)
	at com.google.common.eventbus.Subscriber$SynchronizedSubscriber.invokeSubscriberMethod(Subscriber.java:150)
	at com.google.common.eventbus.Subscriber$1.run(Subscriber.java:76)
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:399)
	at com.google.common.eventbus.Subscriber.dispatchEvent(Subscriber.java:71)
	at com.google.common.eventbus.Dispatcher$PerThreadQueuedDispatcher.dispatch(Dispatcher.java:116)
	at com.google.common.eventbus.EventBus.post(EventBus.java:217)
	at net.minecraftforge.fml.common.LoadController.distributeStateMessage(LoadController.java:148)
	at net.minecraftforge.fml.common.Loader.loadMods(Loader.java:568)
	at net.minecraftforge.fml.client.FMLClientHandler.beginMinecraftLoading(FMLClientHandler.java:231)
	at net.minecraft.client.Minecraft.func_71384_a(Minecraft.java:466)
	at net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:377)
	at net.minecraft.client.main.Main.main(SourceFile:123)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:196)
	at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:231)
	at org.multimc.EntryPoint.listen(EntryPoint.java:143)
	at org.multimc.EntryPoint.main(EntryPoint.java:34)
Caused by: java.lang.IllegalArgumentException: object is not an instance of declaring class
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at net.minecraftforge.fml.common.ILanguageAdapter.lambda$createEventHandler$1(ILanguageAdapter.java:70)
	at net.minecraftforge.fml.common.FMLModContainer.lambda$handleModStateEvent$0(FMLModContainer.java:607)
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.util.stream.AbstractPipeline.evaluateToArrayNode(Unknown Source)
	at java.util.stream.ReferencePipeline.toArray(Unknown Source)
	at net.minecraftforge.fml.common.FMLModContainer.handleModStateEvent(FMLModContainer.java:608)
	... 44 more
```

This change is not currently compatible with existing mods.

Haringat commented:
@mezz which pack exactly did you try? It could be that CraftTweaker is doing something unexpected. Have you tried taking that one out?

liach commented on src/main/java/net/minecraftforge/fml/common/ILanguageAdapter.java:
```diff
@@ -22,17 +22,63 @@
+                }
+                else
+                {
+                    method.invoke(event);
```
The invocation needs a mod instance, I think

covers1624 replied:
That is indeed the case, the event surely isn't the instance we are invoking the method on.

LexManos commented:
I am still not fully understanding what the point of this is. First off, MOST people don't understand that the event handler can be on multiple methods of the same event type. This is a somewhat undeclared/undefined behavior.
Secondly, the only benefit would be from defining the event handler in a way that specifically supports asynchronicity.
Honestly, I think the main bulk of this PR is not needed, nore would be functionality acceptable in Minecraft/Existing environment.
However, one thing that MAY be a reasonable system would be to change eventHandlers from a multimap to a Map<Class<? extends FMLEvent>, Consumer<? extends FMLEvent>>. And at the end of gatherAnnotations() you build that single consumer for each event handler by invoking the language adapter.
And java's implementation would just be the loop that exists today.
That way the Language Adapter defines if multiple methods are innately sync or async. 

But ya, the main concept of 'fixing' the 'issue' of how Forge calls lifecycle event functions I think is not needed to be addressed. First of each function is only ever invoked once at loader per event. The cost is negligible in relation to the implementation of the method itself. I think you're trying to solve this on the wrong side of the method call.
And on top of all that, your PR doesn't actually work and you've admitted yourself that you haven't actually tested it....

mezz commented:
@Haringat unfortunately not public packs. I'll test some more out to see what I can find.
I'm interested to know if you have tested this against packs though. I'm happy to help out and review but you should do the initial testing.

mezz commented:
Here is a crash report from running it with [Modern Skyblock 2](https://minecraft.curseforge.com/projects/modern-skyblock-2/files/2509014): https://gist.github.com/mezz/b10e8262153454cffb465769770fb31b

Haringat committed:
fix handler invokation target and bump forge version (API change)

Haringat committed:
merge

Haringat commented:
Sorry for going silent for so long. I've been quite busy the last weeks.
However, today I've looked at everything and pushed a new build. (And actually tested it with the Modern Skyblocks 2 pack @mezz mentioned)
It should-work™ now. The only mod that caused problems was extra-bees, but that was kinda understandable, as the crash was a NoSuchFieldException: eventMethods (I've changed the type and renamed that to eventHandlers to better reflect what it actually is). I would recommend to leave it at that and not rename the field back because since the type was changed a NoSuchFieldException is more clear telling "Hey, the API changed" than an InvalidCastException that would otherwise be thrown. And since it is and has always been a private field I don't consider that to actually be a bug in my code, but rather a misuse in whoever-wrote-extrabees' code.
I also bumped the API level to make it even more clear. (And of course also because I had to due to semantic versioning/API change in ILanguageAdapter)

Edit:
@LexManos 
> MOST people don't understand that the event handler can be on multiple methods of the same event type.

That's not the main point of this but rather a nice side effect.
> the only benefit would be from defining the event handler in a way that specifically supports asynchronicity.

That may be true, but in my experience whenever you expose a more liberal API someone comes along and completely blows your mind with some slick utilization of that API which you've never even thought about.
> change eventHandlers from a multimap to a Map<Class<? extends FMLEvent>, Consumer<? extends FMLEvent>>. And at the end of gatherAnnotations() you build that single consumer for each event handler by invoking the language adapter.

That's pretty much what I did.

I think you have misunderstood what I am trying to achieve here. This is not mainly for parallelism of multiple event handlers of the same type (though as mentioned above that comes with it) but rather enable parallelism within the event handlers to boost performence. For instance, when testing my build with the modpack it took over 15 minutes to load on my AMD X6 1090T with 8GiB RAM. Of course, throwing more/better hardware at it is one part of the solution but optimizing the code is the other one. And that's what I'm trying here.

mezz commented:
As a rule, we do not break binary compatibility except on new Minecraft versions except in pretty extreme circumstances where it's required. We can't accept a PR that bumps the Forge revision number.

Your changes to the name and type of `eventMethods` are fine though, as you said it's a private field. Anyone reflecting into that pretty much deserves what they get. I'm not sure you need to bump the version at all.

> That may be true, but in my experience whenever you expose a more liberal API someone comes along and completely blows your mind with some slick utilization of that API which you've never even thought about.

I recommend not bringing that up as a positive point, you are describing the bane of Forge's existence.

To make your goal more clear I think it would be helpful if you could answer the following high level questions:
 * What will be the immediate effect of this PR on modpack loading?
 * What optimization does this offer to modpack loading?
 * Is the optimization opt-in, opt-out, or does it always affect everything?
 * What further actions (if any) need to be taken to unlock the potential of this optimization?

Based on the code I have a reasonable idea of the answers here but I want to make sure we're on the same page and that the code is doing what it's supposed to. Thanks!

In the meantime I will go fix extra bees, it was ported by someone else and I should have looked it over before accepting the changes.

Haringat commented:
@mezz I think it would be more sensible to answer your questions on a "mod" level rather than a "modpack" level.
* What will be the immediate effects of this PR on mod loading?
For users: none. If mod language adapter and mod utilize it it should load a bit faster though.
For language adapter maintainers: You can now expose an API for supporting asynchronous event handlers.
For modders: If the adapter of the language you use supports it you can now choose to write your handlers synchronously to work more efficiently.
* What optimization does this offer to mod loading?
None. Though it has the potential to offer optimization by smartly implementing the language adapter. I've provided a sample implementation supporting Futures. But no optimization is done to existing mods.
* Is the optimization opt-in, opt-out or does it affect everything?
For modders: It's opt-in for now as the language adapters support the current way of implementing an event handler (returning a void) as well as another way returning a Future.
For language adapter maintainers: It's opt-out because ILanguageAdapter provides a default implementation. To not support asynchronous handlers you need to override the method in your language adapter in a fitting way.
* What further actions (if any) need to be taken to unlock the potential of this optimization?
For language adapter maintainers: support async as good as you can. Detecting the return type Future is certainly one way but I'm sure there are more sophisticated methods.
For modders: write your handlers asynchronously wherever it seems sensible and return a future that resolves as soon as the work is done. (As mentioned above there might be more to come but that's it for now)

mezz commented:
I don't see any reference to `eventMethods` in Extra Bees, can you send me the crash report?

Your answers look good, I have a follow up question:
 * What does a java mod need to do to take advantage of this?

The only real issue I see here is that cpw is rewriting fml for 1.13 right now, and I have recently learned that he is keeping things like loading parallelism in mind. If we merge the system you have here, it may not see much use because it is going to be replaced within a year.

Haringat commented:
@Mezz Sorry, it wasn't Extra Bees, but Magic Bees that caused problems.
As for the report (if it's still important): https://gist.github.com/Haringat/473164bb97e48da2f00aa8be2921dc8c

> I'm not sure you need to bump the version at all.

I have to because the public API of ILanguageAdapter got a new symbol (public BiFunction<Object, FMLEvent, CompletableFuture<Void>> createEventHandler(Method))

* What does a java mod need to do to take advantage of this?
> write your handlers asynchronously wherever it seems sensible and return a future that resolves as soon as the work is done.

mezz commented:
The new method in `ILanguageAdapter` has a default implementation, it is not a breaking change.

Can you create an example java mod that shows this feature off?
Test mods are located in `src\test\java\net\minecraftforge\debug`

Haringat committed:
added example mod

Haringat commented:
> it is not a breaking change

That's correct and the reason why we don't need a major version bump. But a minor version bump is (according to semantic versioning and the comment right above the changed number) still needed because it's a backwards-compatible API change (in semantic versioning a minor version incrementation).

> Can you create an example java mod that shows this feature off?

Done now. It's inside "net.minecraftforge.debug.AsynchronousEventHandlerTest"

mezz commented:
In practice, the minor version is only bumped for new "recommended" builds. This project does not properly follow semantic versioning.

liach commented on src/test/java/net/minecraftforge/debug/AsynchronousEventHandlerTest.java:
```diff
@@ -0,0 +1,67 @@
+    public Future<Void> asynchronousPreInitializationEventHandler(FMLPreInitializationEvent e)
+    {
+        // This won't block the thread and will run in parallel to the other event handler
+        CompletableFuture[] futures = Stream.of(1,2,3,4,5,6,7,8,9,10)
```
`IntStream.range(0, 10).map()` looks better

Haringat committed:
used IntegerStream.range().boxed() and refactored synchronous handler to show the difference

Haringat committed:
revert Forge API version to 1

Haringat committed:
cleaned up unused import

Haringat commented:
> The only real issue I see here is that cpw is rewriting fml for 1.13 right now, and I have recently learned that he is keeping things like loading parallelism in mind. If we merge the system you have here, it may not see much use because it is going to be replaced within a year.

I don't know how cpw has planned his implementation but I would recommend putting it into the language adapter, especially if he plans on supporting parallelism, because as mentioned (far) above, other languages (e.g. Kotlin) have ways to support parallelism which would be a hassle to support in Java (coroutines). On the other hand writing it in Kotlin would create a hard-dependency on at least 3 external libraries, which is just unnecessary.

0xilly commented:
I would hold off till we see what happens with the new Language Adapter API with the rewrite. Maybe wait till we know what the API looks like and adapt it to that instead of throwing it out with the rewrite.

mezz commented:
Please update this to resolve the (probably minor) conflicts in `ILanguageAdapter`

Commit:
fix merge conflict

Haringat commented:
@mezz I updated it.

LexManos commented:
These events are going away and being re-done completely for 1.13. And I honestly don't think the very little benefit you've described is worth the can of worms this would open.
